### DISCUSSION TITLE
How to use repeater with Spatie Media Library properly

### DISCUSSION THREAD

**Discussion by IgordeOliveira** [filamentphp/filament#10666] [Help]
Hi everyone, I've been spending a lot of time trying to get file input to work within a repeater. I managed to get it to work with functional behavior by creating a UUID to pass as the name of the collection. This prevents the images from mixing together. However, I'm still facing a problem when trying to delete images. When I delete a block from the repeater, it deletes the corresponding index in the JSON column. Unfortunately, the images have no relation to this JSON, so it has no effect. I tried to listen to the model's updating event to delete based on the JSON values of the column that the repeater saves. Unfortunately, it changes the index, making it difficult to do the diff correctly in all cases. Does anyone have any suggestions? My final goal is to upload photos with a description for each photo or in this case, a group of four photos.

`

```
return Tabs\Tab::make('Fotos')
            ->icon('heroicon-s-photo')
            ->schema([
                Forms\Components\Repeater::make('grupo_fotos')
                    ->label('')
                    ->reorderable(false)
                    ->defaultItems(1)
                    ->addActionLabel('Adicionar novo grupo de fotos')
                    ->grid(2)
                    ->schema(function () {
                        return [
                            SpatieMediaLibraryFileUpload::make('fotos')
                                ->multiple()
                                ->image()
                                ->minFiles(1)
                                ->maxFiles(4)
                                ->helperText('Máximo de 4 imagens por grupo')
                                ->panelLayout(null)
                                ->previewable()
                                ->columns(2)
                                ->openable()
                                ->collection(function (Get $get) {
                                    return 'grupo_'.$get('id_for_media');
                                }),
                            Forms\Components\Hidden::make('id_for_media')->default(Str::uuid()->toString()),
                            Forms\Components\TextInput::make('descricao')
                                ->label('Descrição')
                                ->required(),
                        ];
                    }),
            ]
            )
```

;`

**Comment by DanielGamboa** [comment]
There are several options depending on what you want to do.  I'm actually working on this same problem but i feel that I'm close.  My approach is as fallows I have a Clients model  where I set up a hasMany relationship with a pivot table ClientDocuments and the corresponding belongsTo Clients.

I set the hasMedia, interacts with media in ClientDocuments
On the pivot table i have: ID, client_id, doc_type.

With out repeater it works perfectly. Spaties media library grabs ClientDocuments pivot as a model and does its thing.  The problem im having is in how the information is stored in the database.  Out of the box what it does is it leaves media_id on media table empty because of how Filament saves the data to ClientDocuments.
[https://filamentphp.com/docs/3.x/forms/advanced#saving-data-to-a-belongsto-relationship](https://filamentphp.com/docs/3.x/forms/advanced#saving-data-to-a-belongsto-relationship)

My pivot tables new record ID is not set by the time Spatie saves the data to the media table.  So I get all fields filled but media_id remains empty.  (i set it to nullable in while working on this).
If i let it error out the ClientDocuments  is created with the text field, in your case "descricao" so when i go back from the UI i can then attach the media because the ID for ClientDocuments already exists and it saves correctly to the Media table media_id column. After that i can delet or edit just fine.

its not a solution and if i solve it ill be more than happy to pass along the info.  Hope it helps

**Comment by IgordeOliveira** [comment]
Today we have this feature merged, so probally its be the best option to the most cases: https://github.com/filamentphp/filament/pull/10795
